### ISSUE TITLE
Better shell scriping for Titan as a Service -- remove the requirements for multiple shells.

### ISSUE THREAD

**Issue by okram** [thinkaurelius/titan#263]
I think we should make titan.sh more of this nature. Such a script would greatly improve developer/user experience.

```
$ titan start
$ titan stop
$ titan purge
Are you sure you want to clear your graph and start again? [y/N]
```

Below is an example script provided by Harish that wraps the current titan.sh and does just that. I assume that we should have environmental variables such as:

```
$TITAN_HOME
$ELASTICSEARCH_HOME
$CASSANDRA_HOME
$TITAN_CONFIG // like Hadoop, uses the .yaml, .properties, .xml files here for config
```

``` text
#!/bin/bash

NO_ARGS=0
E_OPTERROR=65

if [ $# -eq "$NO_ARGS" ] 
then
        echo "Usage: ./egosystem.sh start|stop|purge"
        exit $E_OPTERROR
fi


function stopEgosystem() {
        echo "Stopping Titan/Cassandra..."
        ps aux | grep cassandra | awk '{print $2}' | xargs sudo kill
        echo "Stopping ElasticSearch..."
        ps aux | grep elasticsearch | awk '{print $2}' | xargs sudo kill
        echo "Done"
}

function startEgoSystem() {
        echo "Starting ElasticSearch..."
        cd /usr/local/elasticsearch-0.90.0/
        sudo nohup bin/elasticsearch -f -Xms15g -Xmx15g config/elasticsearch.yml > /var/data/egosystem/logs/elasticsearch.log &

        sleep 5
        echo "Starting Titan/Cassandra..."
        cd /home/web/titan/
        sudo nohup bin/titan.sh config/titan-server-rexster.xml config/titan-server-cassandra.properties > /var/data/egosystem/logs/titan.log &
        echo "Done"
}

function purgeEgoSystem() {
        echo "Purging data directories..."
        sudo rm -rf /var/data/egosystem/cassandra
        sudo rm -rf /var/data/egosystem/elasticsearch
        echo "Done"
}

if [ "$1" == "start" ]
then
        stopEgosystem
        startEgoSystem
elif [ "$1" == "stop" ]
then
        stopEgosystem
elif [ "$1" == "purge" ]
then
        stopEgosystem
        purgeEgoSystem
        startEgoSystem
        echo "Waiting for titan/rexster to start..."
        sleep 7
        echo "Configuring Schema..."
        bin/gremlin.sh -e /home/web/egosystem/config/Schema.groovy 
fi
```

**Comment by okram** [comment]
To extend why this is important for user experience. Right now, to use Titan, you have to go through the following inane series of steps.
- Open a terminal and start Cassandra.
- Open a terminal and start ElasticSearch.
- Open a terminal and start Rexster
  ...if you are developing and what to purge you data easily.
- CNTRL-C the terminal (or ps aux | grep | kill) and then sudo rm -rf /var/lib/cassandra
- CNTRL-C the terminal (or ps aux | grep | kill) and then sudo rm -rf /var/lib/elasticsearch
- Then restart each service

So painful... it hurts.

**Comment by okram** [comment]
@dalaro wrote: I like it.  By the way, have you or a client needed init.d scripts or systemd service files to automatically start and stop Titan, Rexster, Cassandra, HBase, etc. when the host power cycles?  I think the developer-friendly script called for by the ticket is separate from init.d/systemd, but the two concerns are still potentially related

@okram writes: That would be smart too. I'm not too keen on all the particulars of when Linux loads up and when shells get called here and there, but yea --- the general idea would be to make Titan like MySQL in that it "just exists." Moreover, Harish copied the pattern from Apache web server where he simply does "apachectl start" (stop) when he edits a configuration. He was saying, with so many times he is tweaking cassandra.yaml, elasticsearch.yaml, restarting Rexster (with his new project .jar in classpath), having a simple "titan start" would be very helpful. ... His script is specific to his computer so some form of (best I can think of) $TITAN_CONF-style script would be great. Also, I like his "titan purge" which is good for developers :D.

**Comment by rustyrazorblade** [comment]
I think getting debian packages & rpms would be helpful as well.  Take a
look at https://github.com/tcurdt/jdeb

On Mon, May 6, 2013 at 3:28 PM, Marko A. Rodriguez <notifications@github.com

> wrote:
> 
> @dalaro https://github.com/dalaro wrote: I like it. By the way, have
> you or a client needed init.d scripts or systemd service files to
> automatically start and stop Titan, Rexster, Cassandra, HBase, etc. when
> the host power cycles? I think the developer-friendly script called for by
> the ticket is separate from init.d/systemd, but the two concerns are still
> potentially related
> 
> @okram https://github.com/okram writes: That would be smart too. I'm
> not too keen on all the particulars of when Linux loads up and when shells
> get called here and there, but yea --- the general idea would be to make
> Titan like MySQL in that it "just exists." Moreover, Harish copied the
> pattern from Apache web server where he simply does "apachectl start"
> (stop) when he edits a configuration. He was saying, with so many times he
> is tweaking cassandra.yaml, elasticsearch.yaml, restarting Rexster (with
> his new project .jar in classpath), having a simple "titan start" would be
> very helpful. ... His script is specific to his computer so some form of
> (best I can think of) $TITAN_CONF-style script would be great. Also, I like
> his "titan purge" which is good for developers :D.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/263#issuecomment-17512349
> .

## 

Jon Haddad
http://www.rustyrazorblade.com
skype: rustyrazorblade

**Comment by okram** [comment]
@hariharshankar is pushing on this.
_support_

**Comment by dalaro** [comment]
@hariharshankar: on what Linux distro are you running Titan?

I think that we could write a small main class that, after interactively warning the user and prompting for positive confirmation, calls clearStorage() on Titan's configured index and data backends.  This main class should help the developer use case.  When used with a configuration that embeds Cassandra and ES in Titan's JVM, there would be no need to kill background processes or manually delete directories.

We could also use this purge-main-class in a Titan init.d script.  I think the packaging milestones are (roughly):
1. write an init.d script with start, stop, status, purge (purging is implemented using the main class)
2. build an OS package for Titan (e.g. with jdeb)
3. host OS packages as ordinary file downloads (not a package repo)
4. (maybe) cryptographically sign and host OS packages in a deb/yum repo

The Debian and Fedora packaging toolchains and runtime environments are so different that I would expect minimal if any shared packaging-related work.

**Comment by okram** [comment]
My two cents and take with a grain of salt as I'm not a "devops" guy.

I see this ticket as simply making it easy to start/stop/status Titan (purge optional -- and can be ignored in the initial pass). This script simply starts up Cassandra and ElasticSearch and wa'la. Nothing fancy, just a simple shell script so people don't have multiple terminals open with:

```
cassandra -f
elasticsearch -f
CNTRL-C, CNTRL-C
…
```

**Comment by hariharshankar** [comment]
We are using RedHat in our servers. 
As @okram mentioned, purge is a luxury to have, but would like for start/stop functionalities. The current script works for all our purposes, but the main draw back is that the stop function kills ALL the processes that has the word cassandra in it. This obviously has unintended consequences.

**Comment by mbroecheler** [comment]
This needs to be subsumed by a more general Titan manger deployment service.

**Comment by zcox** [comment]
I know this ticket is closed, but fwiw here is the Upstart script we're using with Titan Server (Cassandra) 0.3.1 on Ubuntu 12.04 on ec2. I literally just used wget to download the .tar.gz and extracted it into /home/ubuntu/titan. Then this seems to work well:

```
author "Zach Cox <zcox@pongr.com>"
description "Titan+Cassandra+Rexster"
version "0.3.1"

#2=regular running mode, 3,4,5 are aliases for 2
start on runlevel [2345]

#0=shutdown, 1=single-user mode, 6=reboot
stop on runlevel [016]

#logs to /var/log/upstart/titan.log
console log

#auto-restarts app if it dies unexpectedly, won't restart if it's stopped properly
respawn

env titan="bin/titan.sh"
env rexster_config="config/titan-server-rexster.xml"
env titan_config="config/titan-server-cassandra-es.properties"

chdir /home/ubuntu/titan

exec $titan $rexster_config $titan_config

#by default, upstart will send the app SIGTERM on stop
```

The above goes in /etc/init/titan.conf and allows us to use the start and stop commands to run Titan Server as a daemon (as well as auto-starts titan if server is rebooted):

```
$ sudo start titan
$ sudo stop titan
```

It has worked well so far during initial testing, but I can't say it's 100% bulletproof.

I also think providing pre-built packages for Debian (and other) systems would help adoption as well as operations.

**Comment by darKoram** [comment]
I downloaded the latest zip of titan-cassandra which doesn't come with the egosystem.sh posted above.  egosystem was EXACTLY what i was looking for.  I had to make a small change to get it to work on a mac.  Without this, the sudo kill would fail if it didn't find a process.  Also, it is tricky to get environment variables set in .bashrc, for example, to pass on to sudo commands.  In the end, i had to make an alias in .bashrc like so:

alias sego="sudo TITAN_HOME=$TITAN_HOME CASSANDRA_HOME=$CASSANDRA_HOME ELASTICSEARCH_HOME=$ELASTICSEARCH_HOME ./egosystem.sh"
and then call like this:
$ sego start

function stopEgosystem() {
        echo "Stopping Titan/Cassandra..."
        result=$(ps aux | grep cassandra | awk '{print $2}' | xargs)
        echo $result
        if [ "$result" != "" ]
        then
          sudo kill $result
        fi
        echo "Stopping ElasticSearch..."
        result=$(ps aux | grep elasticsearch | awk '{print $2}' | xargs)
        if [ "$result" != "" ]
        then
          sudo kill $result
        fi
        echo "Done"
}
